File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/banner/application/BannerUseCase.java
package b1nd.dodam.restapi.banner.application;

import b1nd.dodam.domain.rds.banner.entity.Banner;
import b1nd.dodam.domain.rds.banner.enumeration.BannerStatus;
import b1nd.dodam.domain.rds.banner.repository.BannerRepository;
import b1nd.dodam.restapi.banner.application.data.req.BannerReq;
import b1nd.dodam.restapi.support.data.Response;
import b1nd.dodam.restapi.support.data.ResponseData;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Component
@Transactional(rollbackFor = Exception.class)
@RequiredArgsConstructor
public class BannerUseCase {

    private final BannerRepository repository;

    public Response create(BannerReq req) {
        repository.save(req.mapToBanner());
        return Response.created("배너 생성 성공");
    }

    public Response modify(int id, BannerReq req) {
        Banner banner = repository.getById(id);
        banner.updateBanner(req.title(), req.image(), req.url(), req.expireAt());
        return Response.noContent("배너 수정 성공");
    }

    public Response activate(int id) {
        Banner banner = repository.getById(id);
        banner.activateStatus();
        return Response.noContent("배너 활성화 성공");
    }

    public Response deactivate(int id) {
        Banner banner = repository.getById(id);
        banner.deactivateStatus();
        return Response.noContent("배너 비활성화 성공");
    }

    public Response deleteById(int id) {
        repository.deleteById(id);
        return Response.noContent("배너 삭제 성공");
    }

    @Transactional(readOnly = true)
    public ResponseData<Banner> getById(int id) {
        return ResponseData.ok("배너 단일 조회 성공", repository.getById(id));
    }

    @Transactional(readOnly = true)
    public ResponseData<List<Banner>> getActivates() {
        return ResponseData.ok("활성화 배너 조회 성공", repository.findByStatusOrderByIdDesc(BannerStatus.ACTIVE));
    }

    @Transactional(readOnly = true)
    public ResponseData<List<Banner>> getAllOrderByIdDesc() {
        return ResponseData.ok("배너 전체 조회 성공", repository.findAllByOrderByIdDesc());
    }

}

File: services/service-banner/src/main/kotlin/com/b1nd/dodamdodam/banner/application/banner/BannerUseCase.kt
package com.b1nd.dodamdodam.banner.application.banner

import com.b1nd.dodamdodam.banner.application.banner.data.request.BannerRequest
import com.b1nd.dodamdodam.banner.application.banner.data.response.BannerResponse
import com.b1nd.dodamdodam.banner.application.banner.data.toBannerEntity
import com.b1nd.dodamdodam.banner.domain.banner.service.BannerService
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Transactional(rollbackFor = [Exception::class])
class BannerUseCase(
    private val bannerService: BannerService,
) {
    fun create(request: BannerRequest) {
        bannerService.create(request.toBannerEntity())
    }

    fun modify(id: Long, request: BannerRequest) {
        val banner = bannerService.findById(id)
        banner.updateBanner(
            title = request.title,
            imageUrl = request.image,
            redirectUrl = request.url,
            expireAt = request.expireAt,
        )
    }

    fun activate(id: Long) {
        val banner = bannerService.findById(id)
        banner.activateStatus()
    }

    fun deactivate(id: Long) {
        val banner = bannerService.findById(id)
        banner.deactivateStatus()
    }

    fun deleteById(id: Long) {
        bannerService.deleteById(id)
    }

    @Transactional(readOnly = true)
    fun getById(id: Long): BannerResponse {
        val banner = bannerService.findById(id)
        return BannerResponse.fromEntity(banner)
    }

    @Transactional(readOnly = true)
    fun getActiveBanners(): List<BannerResponse> {
        return bannerService.findActiveBanners().map { BannerResponse.fromEntity(it) }
    }

    @Transactional(readOnly = true)
    fun getAllBanners(): List<BannerResponse> {
        return bannerService.findAll().map { BannerResponse.fromEntity(it) }
    }
}
